fem/tmop/tmop_pa.cpp

```diff
@@ -0,0 +1,347 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "../tmop.hpp"
+#include "../linearform.hpp"
+#include "../pgridfunc.hpp"
+#include "../tmop_tools.hpp"
+#include "../quadinterpolator.hpp"
+#include "../../general/forall.hpp"
+#include "../../linalg/kernels.hpp"
+
+namespace mfem
+{
+
+void TMOP_Integrator::AssembleGradPA(const Vector &xe,
+                                     const FiniteElementSpace &fes)
+{
+   MFEM_VERIFY(PA.R, "PA extension setup has not been done!");
+   PA.setup_Grad = true;
+
+   if (PA.dim == 2)
+   {
+      AssembleGradPA_2D(xe);
+      if (coeff0) { AssembleGradPA_C0_2D(xe); }
+   }
+
+   if (PA.dim == 3)
+   {
+      AssembleGradPA_3D(xe);
+      if (coeff0) { AssembleGradPA_C0_3D(xe); }
+   }
+}
+
+// We might come here w/o knowing that PA will be used.
+// It is the case when EnableLimiting is called before the Setup => AssemblePA.
+void TMOP_Integrator::EnableLimitingPA(const GridFunction &n0)
+{
+   MFEM_VERIFY(PA.enabled, "EnableLimitingPA but PA is not enabled!");
+   const ElementDofOrdering ordering = ElementDofOrdering::LEXICOGRAPHIC;
+
+   // Nodes0
+   const FiniteElementSpace *n0_fes = n0.FESpace();
+   const Operator *n0_R = n0_fes->GetElementRestriction(ordering);
+   PA.X0.SetSize(n0_R->Height(), Device::GetMemoryType());
+   PA.X0.UseDevice(true);
+   n0_R->Mult(n0, PA.X0);
+
+   // Get the 1D maps for the distance FE space.
+   const IntegrationRule &ir = EnergyIntegrationRule(*n0_fes->GetFE(0));
+   PA.maps_lim =
+      &lim_dist->FESpace()->GetFE(0)->GetDofToQuad(ir, DofToQuad::TENSOR);
+
+   // lim_dist & lim_func checks
+   MFEM_VERIFY(lim_dist, "No lim_dist!")
+   const FiniteElementSpace *ld_fes = lim_dist->FESpace();
+   const Operator *ld_R = ld_fes->GetElementRestriction(ordering);
+   MFEM_VERIFY(ld_R, "No lim_dist restriction operator found!");
+   PA.LD.SetSize(ld_R->Height(), Device::GetMemoryType());
```

Comment:
You're right, it should be fine - I'll remove it.

Reply:
Explicitly giving the memory type may be important if later one creates aliases with `some_vec.MakeRef(PA.LD, ...)`. However, this may become unnecessary in the near future with #2203.

Reply:
We'll keep the `Device::GetMemoryType()` then.